[VectorCombine] Fold reduce(trunc(x)) -> trunc(reduce(x)) iff cost effective

RKSimon opened the pull request:
Vector truncations can be pretty expensive, especially on X86, whilst scalar truncations are often free.

If the cost of performing the add/mul/and/or/xor reduction is cheap enough on the pre-truncated type, then avoid the vector truncation entirely.

Fixes https://github.com/llvm/llvm-project/issues/81469

RKSimon committed:
[VectorCombine] Add test coverage for reduce(trunc(X)) -> trunc(reduce(X)) folds

nikic commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1526,6 +1527,67 @@ bool VectorCombine::foldShuffleFromReductions(Instruction &I) {
+    break;
+  default:
+    return false;
+  }
```
Surely this switch already exists somewhere?

RKSimon replied:
Amazingly no, I'll do a cleanup

nikic commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1526,6 +1527,67 @@ bool VectorCombine::foldShuffleFromReductions(Instruction &I) {
+  Value *ReductionSrc = I.getOperand(0);
+
+  Value *TruncSrc;
+  if (!match(ReductionSrc, m_Trunc(m_OneUse(m_Value(TruncSrc)))))
```
Why is there a one-use requirement on the trunc *source*, rather than the trunc? Please add a multi-use test.

RKSimon replied:
My mistake - tests added to check for this

nikic commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1526,6 +1527,67 @@ bool VectorCombine::foldShuffleFromReductions(Instruction &I) {
+    return false;
+
+  auto *Trunc = cast<CastInst>(ReductionSrc);
+  auto *TruncTy = cast<VectorType>(TruncSrc->getType());
```
```suggestion
  auto *TruncSrcTy = cast<VectorType>(TruncSrc->getType());
```
Or `SrcTy`. As is that sounds like the trunc result type.

RKSimon committed:
[VectorCombine] Fold reduce(trunc(x)) -> trunc(reduce(x)) iff cost effective

Vector truncations can be pretty expensive, especially on X86, whilst scalar truncations are often free.

If the cost of performing the add/mul/and/or/xor reduction is cheap enough on the pre-truncated type, then avoid the vector truncation entirely.

Fixes #81469

phoebewang commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -29,6 +29,7 @@
 #include "llvm/IR/PatternMatch.h"
 #include "llvm/Support/CommandLine.h"
 #include "llvm/Transforms/Utils/Local.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
```
Is this required? I don't find anything related to LoopUtils.

nikic replied:
getArithmeticReductionInstruction() is defined there.

RKSimon replied:
Yes, the reduction intrinsics mapping helpers are all defined in LoopUtils.h

nikic reviewed (approved):
LG

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -1526,6 +1528,60 @@ bool VectorCombine::foldShuffleFromReductions(Instruction &I) {
+                           ReductionSrcTy->getScalarType(),
+                           TTI::CastContextHint::None, CostKind);
+
+  if (OldCost < NewCost || !NewCost.isValid())
```
Nice catch - fixed.
